CRM/CiviGeometry/BAO/Geometry.php

```diff
@@ -433,39 +433,147 @@ public static function wkt2kml($wkt) {
-      foreach ($address_ids as $address_id) {
-        $select = CRM_Utils_SQL_Select::from(self::getTableName() . ' g, civicrm_address ca')
-          ->select("ca.id as address_id, g.id as geometry_id")
-          ->where("ST_Contains(g.geometry, ST_GeomFromText(CONCAT('POINT(', ca.geo_code_2, ' ', ca.geo_code_1, ')'), 4326)) = 1")
-          ->where("g.id = #geometry_id", ['geometry_id' => $geometry_id])
-          ->where("ca.id = #address_id", ['address_id' => $address_id]);
-        $result = CRM_Core_DAO::executeQuery($select->toSQL())->fetchAll();
-        if (!empty($result)) {
-          $results[] = $result[0];
+    $tmpTbl = CRM_Utils_SQL_TempTable::build();
+    if ($params['keep_temp_table']) {
+      $tmpTbl->setDurable();
+    } else {
+      $tmpTbl->setAutodrop(true);
+    }
+    $tmpTbl->createWithColumns('address_id INT');
+
+    $tempTableName = $tmpTbl->getName();
+    if ($params['precheck_relationships']) {
+      CRM_Core_DAO::executeQuery("
+        INSERT INTO $tempTableName (address_id)
+        SELECT ca.id
+        FROM
+          civicrm_address ca
+        WHERE
+              ca.geo_code_2 >= %1
+          AND ca.geo_code_2 <= %2
+          AND ca.geo_code_1 <= %3
+          AND ca.geo_code_1 >= %4
+          AND NOT EXISTS (
+            SELECT *
+            FROM civigeometry_geometry_entity cge
+            WHERE
+              cge.entity_id = ca.id
+              AND cge.geometry_id = %5
+              AND cge.entity_table = 'civicrm_address'
+          )
+        ORDER BY ca.id
+      ", [
+        1 => [$bounds['left_bound'], 'Float'],
+        2 => [$bounds['right_bound'], 'Float'],
+        3 => [$bounds['top_bound'], 'Float'],
+        4 => [$bounds['bottom_bound'], 'Float'],
+        5 => [$geometry_id, 'Integer']
+      ]);
+    } else {
+      CRM_Core_DAO::executeQuery("
+        INSERT INTO $tempTableName (address_id)
+        SELECT ca.id
+        FROM civicrm_address ca
+        WHERE
+              ca.geo_code_2 >= %1
+          AND ca.geo_code_2 <= %2
+          AND ca.geo_code_1 <= %3
+          AND ca.geo_code_1 >= %4
+        ORDER BY ca.id
+      ", [
+        1 => [$bounds['left_bound'], 'Float'],
+        2 => [$bounds['right_bound'], 'Float'],
+        3 => [$bounds['top_bound'], 'Float'],
+        4 => [$bounds['bottom_bound'], 'Float'],
+      ]);
+    }
+
+    $numCandidates = CRM_Core_DAO::singleValueQuery("SELECT COUNT(*) FROM $tempTableName");
+
+    function fetchCandidateBatch($candidateTable, $offset, $batchSize) {
+      $dao = CRM_Core_DAO::executeQuery("
+          SELECT ct.address_id
+          FROM $candidateTable ct
+          ORDER BY address_id
+          LIMIT %1
+          OFFSET %2
+        ", [
+          1 => [$batchSize, 'Integer'],
+          2 => [$offset, 'Integer'],
+        ]);
+
+        $results = $dao->fetchAll();
+
+        return $results;
+    }
+
+    function geometryContainsCandidate($geometryId, $addressId, $geomTableName) {
```

Comment:
Same comment as L519

Reply:
Also not sure why this needs to be nested within another function probably easiest just to be a function on its own right?